File: fractal-viewer/c/src/viewport.c
#include "viewport.h"

void viewport_init(Viewport *vp, int width, int height) {
    vp->width = width;
    vp->height = height;
    vp->cx = -0.5;
    vp->cy = 0.0;
    vp->zoom = 200.0;
}

void viewport_screen_to_complex(Viewport *vp, int sx, int sy, double *px, double *py) {
    *px = (sx - vp->width / 2.0) / vp->zoom + vp->cx;
    *py = (sy - vp->height / 2.0) / vp->zoom + vp->cy;
}

void viewport_zoom_at(Viewport *vp, int sx, int sy, double factor) {
    double px, py;
    viewport_screen_to_complex(vp, sx, sy, &px, &py);
    vp->zoom *= factor;
    vp->cx = px - (sx - vp->width / 2.0) / vp->zoom;
    vp->cy = py - (sy - vp->height / 2.0) / vp->zoom;
}

void viewport_pan(Viewport *vp, double dx, double dy) {
    vp->cx -= dx / vp->zoom;
    vp->cy -= dy / vp->zoom;
}

File: fractal-viewer/python/viewport.py
class Viewport:
    def __init__(self, width, height):
        self.width = width
        self.height = height
        self.cx = -0.5
        self.cy = 0.0
        self.zoom = 200.0

    def screen_to_complex(self, sx, sy):
        px = (sx - self.width / 2) / self.zoom + self.cx
        py = (sy - self.height / 2) / self.zoom + self.cy
        return px, py

    def zoom_at(self, sx, sy, factor):
        px, py = self.screen_to_complex(sx, sy)
        self.zoom *= factor
        self.cx = px - (sx - self.width / 2) / self.zoom
        self.cy = py - (sy - self.height / 2) / self.zoom

    def pan(self, dx, dy):
        self.cx -= dx / self.zoom
        self.cy -= dy / self.zoom
